### PR TITLE
Feature/repository

### PR DESCRIPTION
**Pull request by qejk**

### CODE REVIEW

File: packages/app/source/server/projections/todos-projection.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 Space.eventSourcing.Projection.extend('Todos.TodosProjection', {
 
-  collections: {
-    todos: 'Todos.TodoLists'
+  dependencies: {
```
Comment (rhyslbw) [inline packages/app/source/server/projections/todos-projection.js:3]: Are you considering projection rebuilding out of scope for this initial phase?

File: packages/app/source/server/projections/todos-projection.js [hunk 2]
```diff
@@ -16,50 +16,31 @@ Space.eventSourcing.Projection.extend('Todos.TodosProjection', {
   },
 
   _onTodoListCreated(event) {
-    this.todos.insert({
-      _id: event.sourceId.toString(),
-      name: event.name,
-      todos: []
-    });
+    this.repository.createTodoList(event.sourceId, event.name);
```
Comment (rhyslbw) [inline packages/app/source/server/projections/todos-projection.js:19]: Repositories here are making projections really expressive and explicit :+1:

**Comment by rhyslbw** [comment]
Awesome work @qejk. Really nice to see a technology agnostic projection

**Comment by rhyslbw** [comment]
Guessing this is still WIP ? Slap the label and assign someone for next action :-)

**Comment by darko-mijic** [comment]
@qejk What else do you think needs to be done to finish this?

**Comment by qejk** [comment]
Would be nice to use new serializable feature for Entities (`Todos.TodoItem`) and change that in `todo-list.js` so creating new instance of of `Todos.TodoItem` would happen on command handler - not in event handler.

This will shine later on on repository - `todo.toData()`, however this could be done on separate branch right after this one since it will cascade on tests.

This pattern kind of created a question in my head - shouldn't we pass empty todos on list creation **in event** - and not do that directly on event handler?

```
  _onTodoListCreated(event) {
    this._assignFields(event);
    this.todos = [];
  },
```

If that question is outside example and more about pattern: I personally will tinker around moving the current logic in `Repository` in my projects that is related to mongo implementation - to separate Data Access Object's (`DAO`) related only to mongo. And then use them as dependency on `Repository`. Should bring more sanity if multiple db types are used.

**Comment by qejk** [comment]
Also since I still don't know how projection rebuilding yet work, considerate:

> Are you considering projection rebuilding out of scope for this initial phase?
> 
> > ah right, yer the projection class would need to be upgraded

by @rhyslbw

File: packages/app/source/server/projections/todos-projection.js (hunk 1 above)
Comment (rhyslbw) [inline packages/app/source/server/projections/todos-projection.js:3]: @qejk If you move this into the `collections` object I believe any repositories backed by `Mongo.Collection`s will work without any changes as `ProjectionRebuilder` [uses this API](https://github.com/meteor-space/event-sourcing/blob/master/source/server/infrastructure/projection-rebuilder.coffee#L54). We could rename it but `collections` is generic enough in my mind, however we could make a backwards compatible change and depreciate if not.

To support other persistence technologies we will need to flesh out the projection repository api to support the rebuilding calls in [rebuild](https://github.com/meteor-space/event-sourcing/blob/master/source/server/infrastructure/projection-rebuilder.coffee#L11-L47)

Comment (rhyslbw) [inline packages/app/source/server/projections/todos-projection.js:3]: … which I just realised there is no abstract `ProjectionRepository` class, but the point still stands in that the `ProjectionRebuilder` should just be calling a method to coordinate the rebuild, therefore making it possible to support any technology

Comment (rhyslbw) [inline packages/app/source/server/projections/todos-projection.js:3]: Ok, no we'll need to make some changes as it's more hard wired than I thought. It's time to review how `ProjectionRebuilder` can be reworked

**Pull request closed without merging**